### PR TITLE
Initial changes to move the server architecture to a layered approach.

### PR DESCRIPTION
**Pull request by DanielBunting**
This an initial change to allow testing,

My idea, was to go for flat controllers, to minimise the testing needed at that level, and have all of the 'business logic' seperated into a manager layer.

It would make sense to add a 'store' layer or similar to seperate the persistence, to allow the testing to focus only on the business logic for the managers, rather than needing to plumb through a system that needs access to an in-memory EF instance for testing. 

Also - sorry for the unsolicited PR, currently in China behind the great firewall, so I'm struggling to verify anything that needs access to my Gmail.

### CODE REVIEW

**Comment by enkodellc** [comment]
From Daniel: 

> For this approach, I'm trying to go with the following:
> 
> Flat Controllers - the least business logic possible in the controllers, allowing us to not need to heavily pepper them with non-integration tets.
> Managers - All of the business logic will live inside of managers, and not much else.
> Store - A seperate project that has the EF dependencies loosely coupled to the managers. IE, we would pass in an instance of IMessageStore to MessageManager instances. The MessageStore instances would have the actual logic required for building the query, again seperating this from business logic

File: src/BlazorBoilerplate.Server/Controllers/EmailController.cs [hunk 1]
```diff
@@ -16,81 +12,24 @@ namespace BlazorBoilerplate.Server.Controllers
...
+        public async Task<ApiResponse> Send(EmailDto parameters) 
+            => ModelState.IsValid ?
+                await _emailManager.Send(parameters) :
+                new ApiResponse(400, "User Model is Invalid");
 
         [HttpGet("Receive")]
         [Authorize]
-        public async Task<ApiResponse> Receive()
-        {
-            //Check email from default account defined in appsettings.json
```
Comment (enkodellc) [inline src/BlazorBoilerplate.Server/Controllers/EmailController.cs]: @DanielBunting  I do like the leaner code with lambda functions but as boilerplate I feel that we need to keep the comments in the code for newbies. You wouldn't believe how many questions I get between BB and MatBlazor.

**Comment by enkodellc** [comment]
@DanielBunting  

- I completely agree on Flat controllers. Definitely needed to be cleaned up. 
- Managers are renamed Services. I named them "Services" form looking at other CLEAN architecture patterns so maybe my terminology was incorrect and could cause confusion.
-  Store, I think I can agree with this as well so then it should make it easier for others to swap out EF for Dapper, other ORM etc.  Is this correct.

In the end this isn't too dramatic and I see where this will make Tests much easier to implement. 

So one final question is are you will to start a testing project as well. If so I hope I can recruit some of these newbies who I have helped to dive in and build some tests to get their feet wet!

**Comment by DanielBunting** [comment]
@enkodellc 

> * I completely agree on Flat controllers. Definitely needed to be cleaned up.

I've made them as flat as I can, where I know how. The ExternalAuthController will need some more thought, as I dodn't want to make changes that were viral in the initial PR. 

> * Managers are renamed Services. I named them "Services" from looking at other CLEAN architecture patterns so maybe my terminology was incorrect, or theirs was.

I don't think either name is wrong, I'm happt to change these all to Services, if you'd prefer. My two cents on this, it Services is a very over-loaded term.. managers is also, but it seems to be the approach MS have taken with UserManager, RoleManager etc.  

> * Store, I think I can agree with this as well so then it should make it easier for others to swap out EF for Dapper, other ORM etc.  Is this correct?

Yep, as long as the interfaces, for the stores are fulfilled, you should be able to swap in anything. 

> So one final question is are you will to start a testing project as well. If so I hope I can recruit some of these newbies who I have helped to dive in and build some tests to get their feet wet!

Yep, that's the plan - After there are seperate stores, and no passing of the DbContext to managers, setting up testing should be quite easy.

**Comment by MarkStega** [comment]
Overall I like this approach. On the server side I'd suggest an addition directory layer to mirror (mostly) the Services in the UI side with Managers\Contracts, Managers\Implementations, and Managers\Controllers.

**Comment by DanielBunting** [comment]
> Overall I like this approach. On the server side I'd suggest an addition directory layer to mirror (mostly) the Services in the UI side with Managers\Contracts, Managers\Implementations, and Managers\Controllers.

I agree with these - Are you happy if I treat this branch as an integration branch, and make a few pull requests into it, and address it in a future PR? I'm trying to keep my changes as isolated as possible atm.

**Comment by enkodellc** [comment]
@DanielBunting  : sounds good.

**Pull request merged**